src/reverse.jl

```diff
@@ -5,105 +5,126 @@
+    nblocks = cld(length(input), nthreads)
 
     @cuda threads=nthreads blocks=nblocks kernel(input, output)
 end
 
-# in-place version, swapping two elements on half the number of threads
-function _reverse(data::AnyCuArray{T, N}; dims::Integer=1) where {T, N}
-    shape = [size(data)...]
-    numelemsinprevdims = prod(shape[1:dims-1])
-    numelemsincurrdim = shape[dims]
+# in-place version for multiple dimensions
+function _reverse!(data::AnyCuArray{T, N}; dims=1:ndims(data)) where {T, N}
+    rev_dims = ntuple((d)-> d in dims && size(data, d) > 1, N)
+    half_dim = findlast(rev_dims)
+    if isnothing(half_dim)
+        # no reverse operation needed at all in this case.
+        return
+    end
+    ref = size(data) .+ 1
+    # converts an ND-index in the data array to the linear index
+    lin_idx = LinearIndices(data)
+    reduced_size = ntuple((d)->ifelse(d==half_dim, cld(size(data,d),2), size(data,d)), N)
+    reduced_length = prod(reduced_size)
+    # converts a linear index in a reduced array to an ND-index, but using the reduced size
+    nd_idx = CartesianIndices(reduced_size)
 
     function kernel(data::AbstractArray{T, N}) where {T, N}
         offset_in = blockDim().x * (blockIdx().x - 1i32)
 
         index_in = offset_in + threadIdx().x
 
-        # the index of an element in the original array along dimension that we will flip
-        #assume(numelemsinprevdims > 0)
-        #assume(numelemsincurrdim > 0)
-        ik = ((cld(index_in, numelemsinprevdims) - 1) % numelemsincurrdim) + 1
-
-        index_out = index_in + (numelemsincurrdim - 2ik + 1) * numelemsinprevdims
-
-        if index_in <= length(data) && index_in < index_out
-            @inbounds begin
-                temp = data[index_out]
-                data[index_out] = data[index_in]
-                data[index_in] = temp
+        if index_in <= reduced_length 
+            idx = Tuple(nd_idx[index_in])
+            index_in = lin_idx[idx...]
+            idx = ifelse.(rev_dims, ref .- idx, idx)
+            index_out =  lin_idx[idx...] 
+
+            if index_in < index_out
+                @inbounds begin
+                    temp = data[index_out]
+                    data[index_out] = data[index_in]
+                    data[index_in] = temp
+                end
             end
         end
 
         return
     end
 
-    # NOTE: we launch twice the number of threads, which is wasteful, but the ND index
-    #       calculations don't allow using only the first half of the threads
-    #       (e.g. [1 2 3; 4 5 6] where threads 1 and 2 swap respectively (1,2) and (2,1)).
+    # NOTE: we launch slightly more than half the number of elements in the array as threads.
+    # The last non-singleton dimension along which to revert is used to define how the array is split.
+    # Only the middle row in case of an odd array dimension could cause trouble, but this is prevented by
+    # ignoring the threads that cross the mid-point
```

Comment:
```suggestion
    # NOTE: we launch slightly more than half the number of elements in the array as threads.
    # The last non-singleton dimension along which to reverse is used to define how the array is split.
    # Only the middle row in case of an odd array dimension could cause trouble, but this is prevented by
    # ignoring the threads that cross the mid-point
```